File: C(++)/004. Median of Two Sorted Arrays.c
/*
004. Median of Two Sorted Arrays
https://leetcode.com/problems/median-of-two-sorted-arrays/

c 19 ms
*/

/*
There are two sorted arrays nums1 and nums2 of size m and n respectively.

Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

Example 1:
nums1 = [1, 3]
nums2 = [2]

The median is 2.0

Example 2:
nums1 = [1, 2]
nums2 = [3, 4]

The median is (2 + 3)/2 = 2.5
*/

int work(int *a1,int l1,int *a2,int l2,int k)
{
    int s1=0,e1=l1-1,s2=0,e2=l2-1,p1=l1/2,p2=l2/2,lsum;
    while (1)
    {
        if (e1-s1+1+e2-s2+1==k) return a1[e1]>a2[e2]?a1[e1]:a2[e2];
        lsum=p1-s1+1+p2-s2+1;
        if (a1[p1]<a2[p2])
        {
            if (lsum>k)
            {
                e2=p2-1;
                p2=s2+(e2-s2+1)/2;
            }
            else if (lsum<=k)
            {
                k-=(p1-s1+1);
                s1=p1+1;
                p1=s1+(e1-s1+1)/2;
            }
        }
        else
        {
            if (lsum>k)
            {
                e1=p1-1;
                p1=s1+(e1-s1+1)/2;
            }
            else if (lsum<=k)
            {
                k-=(p2-s2+1);
                s2=p2+1;
                p2=s2+(e2-s2+1)/2;
            }
        }
        if (s1>e1) return a2[s2+k-1];
        if (s2>e2) return a1[s1+k-1];
    }
}
double direct(int *a,int l)
{
    if (l&1) return (double)a[l/2];
    else return ((double)a[l/2-1]+a[l/2])/2;
}
double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {
    int *a1=nums1,l1=nums1Size,*a2=nums2,l2=nums2Size,n=l1+l2;
    if (l1==0)
        return direct(a2,l2);
    if (l2==0)
        return direct(a1,l1);
    double ans=work(a1,l1,a2,l2,(n+1)/2);
    if (!(n&1))
        ans=(ans+work(a1,l1,a2,l2,n/2+1))/2;
    return ans;
}

File: Java/004. Median of Two Sorted Arrays.java
/*
004. Median of Two Sorted Arrays
https://leetcode.com/problems/median-of-two-sorted-arrays/

java 65 ms
*/

// lsum=p1-st1+1+p2-st2，并且比较时使用lsum<=k-1会TLE
public class Solution {
	public double direct(int temp[])
	{
		int len=temp.length;
		if (len%2==1) return temp[len/2];
		else return (temp[len/2-1]+temp[len/2])/2.;
	}
	public int kthmin(int arr1[],int arr2[],int k)
	{
        int st1=0,ed1=arr1.length-1,p1=arr1.length/2;
        int st2=0,ed2=arr2.length-1,p2=arr2.length/2;
        while (true)
        {
        	// 如果总个数正好为k，那么最大的那个就是
        	if (ed1-st1+1+ed2-st2+1==k)
        		return arr1[ed1]>arr2[ed2]?arr1[ed1]:arr2[ed2];
        	int lsum=p1-st1+1+p2-st2+1; // 有lsum个数比arr2[p2]或arr1[p1]小
        	if (arr1[p1]<arr2[p2])
        	{
        		/* 这里如果lsum==k-1，并不能直接返回arr2[p2]
        		因为arr1[p1+1~ed1]和arr2[p2]的大小关系未知，例如{1,2,3}和{4,5} */
        		
        		// 比arr2[p2]小的数不够k-1个，意味着答案是比arr2[p2]大的数
        		// 等号放在这里，因为arr1[p1+1~ed1]不确定，而确定去除arr1[st1~p1]
        		if (lsum<=k)
        		{
        			k-=p1-st1+1; // 这些数都比arr2[p2]小，以后不关心
        			st1=p1+1;
        			p1=(st1+ed1)/2;
        		}
        		else // 答案在arr1[0~p1]和arr2[0~p2-1]中
        		{
        			ed2=p2-1;
        			p2=(st2+ed2)/2;
        		}
        	}
        	else if (arr1[p1]>=arr2[p2])
        	{
        		if (lsum<=k)
        		{
        			k-=p2-st2+1;
        			st2=p2+1;
        			p2=(st2+ed2)/2;
        		}
        		else
        		{
        			ed1=p1-1;
        			p1=(st1+ed1)/2;
        		}
        	}
        	// 如果其中一个去处完了，那么答案在另一个数组中的位置直接确定
        	if (st1>ed1) return arr2[st2+k-1];
        	if (st2>ed2) return arr1[st1+k-1];
        }
	}
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        if (nums1.length==0)
        	return direct(nums2);
        if (nums2.length==0)
        	return direct(nums1);
        
        int sumlen=nums1.length+nums2.length;
        double ret=kthmin(nums1,nums2,sumlen/2+1)*1.;
        if (sumlen%2==0)
        	ret=(ret+kthmin(nums1,nums2,sumlen/2))/2.;
        return ret;
    }
}
